core/src/main/scala/chisel3/Module.scala

```diff
@@ -271,21 +254,76 @@ abstract class Module extends RawModule {
     }
   }
 
+  // Overridden by OverrideClock
+  private[chisel3] def _implicitClock: Clock = clock
+  // Overridden by OverrideReset
+  private[chisel3] def _implicitReset: Reset = reset
+
   // Setup ClockAndReset
-  Builder.currentClock = Some(clock)
-  Builder.currentReset = Some(reset)
+  Builder.currentClock = Some(_implicitClock)
+  Builder.currentReset = Some(_implicitReset)
   // Note that we do no such setup for disable, it will default to hasBeenReset of the currentReset
   Builder.clearPrefix()
 
   private[chisel3] override def initializeInParent(): Unit = {
     implicit val sourceInfo = UnlocatableSourceInfo
 
     super.initializeInParent()
-    clock := _override_clock.getOrElse(Builder.forcedClock)
-    reset := _override_reset.getOrElse(Builder.forcedReset)
+    clock := Builder.forcedClock
+    reset := Builder.forcedReset
   }
 }
 
+/** Enables overriding of the implicit reset used _within_ the module
+  *
+  * @example {{{
+  * class MyModule extends Module with OverrideReset {
+  *   val gate = IO(Input(Bool()))
+  *   // It it useful to name this signal
+  *   val myReset = reset.asBool || gate
+  *   internalReset := myReset
+  * }
+  * }}}
+  */
+trait OverrideReset { self: Module =>
```

Comment:
must this be a Module? can it work for a RawModule as well?

Reply:
Now it can!